File: week4_tutorial/IPC/python/shared_memory.py
import mmap

size = 1024
key = 1234

# Create a shared memory segment and map it to a writable memory-mapped file
shm = mmap.mmap(-1, size, mmap.MAP_SHARED, mmap.PROT_WRITE, mmap.MAP_ANONYMOUS | mmap.MAP_SHARED, key)

data_to_write = b"Hello from Python!"
shm.write(data_to_write)

print("Data written to shared memory")

# Read data from the shared memory
shm.seek(0)  # Move the file pointer to the beginning of the shared memory
data_read = shm.read(len(data_to_write)).decode("utf-8")

print(f"Data read from shared memory: {data_read}")

# Clean up by closing the memory-mapped file
shm.close()

File: week4_tutorial/IPC/go/shared_memory.go
package main

import (
	"fmt"
	"os"
	"syscall"
	"unsafe"
)

func main() {
	// Specify the key and size for shared memory
	key := int64(1234)
	size := int64(1024)

	// Create a shared memory segment
	shmID, _, err := syscall.Syscall(syscall.SYS_SHMGET, key, size, syscall.IPC_CREAT|0666)
	if err != 0 {
		fmt.Printf("Error creating shared memory segment: %v\n", err)
		os.Exit(1)
	}

	// Attach the shared memory segment to a pointer
	shmPtr, _, err := syscall.Syscall(syscall.SYS_SHMAT, shmID, 0, 0)
	if err != 0 {
		fmt.Printf("Error attaching shared memory segment: %v\n", err)
		os.Exit(1)
	}

	// Data to be written to shared memory
	data := []byte("Hello from Go!")

	// Copy the data into shared memory
	copy((*[1 << 30]byte)(unsafe.Pointer(shmPtr))[:len(data)], data)

	fmt.Println("Data written to shared memory")

	// Clean up by detaching the shared memory segment
	_, _, err = syscall.Syscall(syscall.SYS_SHMDT, shmPtr, 0, 0)
	if err != 0 {
		fmt.Printf("Error detaching shared memory segment: %v\n", err)
		os.Exit(1)
	}
}
